### ISSUE TITLE
Allow concurrent TRUNCATE TABLE operations

### ISSUE THREAD

**Issue by nodep** [scylladb/scylladb#21946]
Currently, TRUNCATE TABLE for tablets does not allow concurrent operations. If a truncate operation is received while another truncate is being executed, the new truncate operation will fail with an error: `Another global topology request is ongoing, please retry.`

While truncate is a rare operation, it should be possible to handle concurrent truncates without error. Putting the incoming truncate operations in a queue (by saving them as separate requests in `system.topology_requests`), while the topology coordinator's truncate handler executes them in sequence is a possible fix.

**Comment by kbr-scylla** [comment]
I think this is basically a duplicate of https://github.com/scylladb/scylladb/issues/16822.

**Comment by avikivity** [comment]
We can say this issue is subordinate to #16822. That one creates a request queue, this one modifies truncate to use it.

**Comment by dani-tweig** [comment]
@nodep @fruch , what needs to be done here?
Will it make 2025.2.0? (the coming few days)

**Comment by nodep** [comment]
This will certainly not make it for 2025.2. This is dependent on #16822 which has a 2025.3 milestone.

**Comment by fruch** [comment]
@nodep now that https://github.com/scylladb/scylladb/issues/16822 is in, we are getting this problem on every single job of rolling upgrade,
can you advise what we should/can do about it ?

here's one example of it:
```
2025-06-28 03:40:22.160: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=4aed9966-3e10-4f7d-9339-30e6cf2d8bc6, source=UpgradeTest.test_generic_cluster_upgrade (upgrade_test.UpgradeTest)() message=Traceback (most recent call last):
File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 970, in test_generic_cluster_upgrade
self._start_and_wait_for_node_upgrade(node_to_upgrade, step=next(step))
File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 877, in _start_and_wait_for_node_upgrade
self.upgrade_node(node, upgrade_sstables=self.params.get('upgrade_sstables'))
File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 81, in inner
self.cql_truncate_simple_tables(session=session, rows=NUMBER_OF_ROWS_FOR_TRUNCATE_TEST)
File "/home/ubuntu/scylla-cluster-tests/sdcm/fill_db_data.py", line 3057, in cql_truncate_simple_tables
session.execute(truncate_query % i)
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/common.py", line 1332, in execute_verbose
return execute_orig(*args, **kwargs)
File "cassandra/cluster.py", line 2802, in cassandra.cluster.Session.execute
File "cassandra/cluster.py", line 5206, in cassandra.cluster.ResponseFuture.result
cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 10.142.15.202:9042 us-east1>: ConnectionException('Host has been marked down or removed'), <Host: 10.142.15.208:9042 us-east1>: <Error from server: code=1003 [Error during truncate] message="Error during truncate: exceptions::invalid_request_exception (Another global topology request is ongoing, please retry.)">, <Host: 10.142.15.210:9042 us-east1>: <Error from server: code=1003 [Error during truncate] message="Error during truncate: exceptions::invalid_request_exception (Another global topology request is ongoing, please retry.)">})
```

## Packages

Base Scylla version: `2025.1.3-20250529.5a67119dce9d` with build-id `0514af4d604db948c535b8a7371221a3c040fb8b`
Target Scylla version (or git commit hash): `2025.3.0~dev-20250627.9d70e7a0674c` with build-id `976aec9516beb8249cfb1fb48a2fa0b4ed848928`




Kernel Version: `6.11.0-1015-gcp`

## Installation details

Cluster size: 3 nodes (n2-highmem-64)

Scylla Nodes used in this run:
- rolling-upgrade-artifacts--ubuntu-n-db-node-4ae4154d-0-3 (34.73.84.162 | 10.142.15.210) (shards: 60)
- rolling-upgrade-artifacts--ubuntu-n-db-node-4ae4154d-0-2 (35.185.116.144 | 10.142.15.208) (shards: 60)
- rolling-upgrade-artifacts--ubuntu-n-db-node-4ae4154d-0-1 (34.75.250.42 | 10.142.15.202) (shards: 60)


OS / Image: `https://www.googleapis.com/compute/v1/projects/ubuntu-os-cloud/global/images/family/ubuntu-2404-lts-amd64` (gce: undefined_region)

Test: `rolling-upgrade-ubuntu2404-test`
Test id: `4ae4154d-a5d3-410f-a9bd-b31fc26776e4`
Test name: `scylla-master/rolling-upgrade/rolling-upgrade-ubuntu2404-test`
Test method: `upgrade_test.UpgradeTest.test_generic_cluster_upgrade`
Test config file(s):

- [generic-rolling-upgrade.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/6485769ceb6543e64e1cfb13c9863f7359b895cd/test-cases/upgrades/generic-rolling-upgrade.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 4ae4154d-a5d3-410f-a9bd-b31fc26776e4`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=4ae4154d-a5d3-410f-a9bd-b31fc26776e4)
- Show all stored logs command: `$ hydra investigate show-logs 4ae4154d-a5d3-410f-a9bd-b31fc26776e4`


## Logs:
- **db-cluster-4ae4154d.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/db-cluster-4ae4154d.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/db-cluster-4ae4154d.tar.zst)
- **schema-logs-4ae4154d.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/schema-logs-4ae4154d.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/schema-logs-4ae4154d.tar.zst)
- **sct-runner-events-4ae4154d.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/sct-runner-events-4ae4154d.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/sct-runner-events-4ae4154d.tar.zst)
- **sct-4ae4154d.log.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/sct-4ae4154d.log.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/sct-4ae4154d.log.tar.zst)
- **loader-set-4ae4154d.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/loader-set-4ae4154d.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/loader-set-4ae4154d.tar.zst)
- **monitor-set-4ae4154d.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/monitor-set-4ae4154d.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/monitor-set-4ae4154d.tar.zst)
- **ssl-conf-4ae4154d.tar.zst** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/ssl-conf-4ae4154d.tar.zst](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/20250628_034141/ssl-conf-4ae4154d.tar.zst)
- **builder-4ae4154d.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/upload_20250628_034319/builder-4ae4154d.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/4ae4154d-a5d3-410f-a9bd-b31fc26776e4/upload_20250628_034319/builder-4ae4154d.log.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/rolling-upgrade/job/rolling-upgrade-ubuntu2404-test/40/)
[Argus](https://argus.scylladb.com/test/e5f0ea7c-9f02-419c-a520-7afd2431e199/runs?additionalRuns[]=4ae4154d-a5d3-410f-a9bd-b31fc26776e4)
</details>

**Comment by mykaul** [comment]
@nodep - status?

**Comment by nodep** [comment]
@mykaul Having the global topology operation queue reduces the problems we had with this. Before, any ongoing TRUNCATE caused an incoming TRUNCATE to fail. With the global queue, a newly received TRUNCATE will be queued, and will not fail. An exception to this is when we receive a TRUNCATE for a table which is currently being truncated (not just queued). The reason for this being that an ongoing TRUNCATE has already started deleting data, but the new TRUNCATE should also delete data which has been received after the ongoing TRUNCATE has started.
One way to get around this would be to also queue a new TRUNCATE for a table which is currently being truncated, instead of failing the new request.

@tgrabiec What do you think?

**Comment by tgrabiec** [comment]
> One way to get around this would be to also queue a new TRUNCATE for a table which is currently being truncated, instead of failing the new request.
> 
> @tgrabiec What do you think?

Makes sense.

**Comment by nodep** [comment]
My previous assessment was wrong.
If the global topology request queue is enabled (via cluster feature flag), every TRUNCATE will be queued, and we don't have a problem with `Another global topology request is ongoing, please retry.`
In cases where some nodes are running versions which don't support the request queue, we fall back on waiting for an existing TRUNCATE for the same table (but only if it is waiting, not if it is already running), or fail the incoming truncate with the `Another global topology request is ongoing`.
I'm afraid there is not much else we can do to improve this for cases where the global topology request queue is not available.
